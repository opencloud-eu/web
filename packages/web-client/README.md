# web-client

The `web-client` is a standalone package that allows you to interact with the [OpenCloud](https://github.com/opencloud-eu/opencloud/) APIs via TypeScript. It provides an abstraction layer between the server and a (web-) application that converts API data into objects with helpful types and utilities. This abstraction ensures that users of the APIs don't need in-depth knowledge about them, such as required methods or returned status codes.

The supported APIs are:

- Graph (drives, sharing, user & group management)
- OCS (capabilities & url signing)
- WebDAV (file operations)

## Installation

Depending on your package manager, run one of the following commands:

```
$ npm install @opencloud-eu/web-client

$ pnpm add @opencloud-eu/web-client

$ yarn add @opencloud-eu/web-client
```

## Usage

### Graph

The graph client needs to be instantiated with a base URI corresponding to your OpenCloud deployment and an axios instance. The axios instance is being used for all requests, which means it needs to include all relevant headers either statically or via interceptor.

```
import axios from axios
import { graph } from '@opencloud-eu/web-client'

const accessToken = 'some_access_token'
const baseURI = 'some_base_uri'

const axiosClient = axios.create({
	headers: { Authorization: accessToken }
})

const graphClient = graph(baseURI, axiosClient)
```

The following example demonstrates how to retrieve all spaces accessible to the user. A `SpaceResource` can then be used to e.g. fetch files and folders (see webdav example down below).

```
const mySpaces = await graphClient.drives.listMyDrives()
```

### OCS

The ocs client needs to be instantiated with a base URI corresponding to your OpenCloud deployment and an axios instance. The axios instance is being used for all requests, which means it needs to include all relevant headers either statically or via interceptor.

```
import axios from axios
import { ocs } from '@opencloud-eu/web-client'

const accessToken = 'some_access_token'
const baseURI = 'some_base_uri'

const axiosClient = axios.create({
	headers: { Authorization: accessToken }
})

const ocsClient = ocs(baseURI, axiosClient)
```

The following examples demonstrate how to fetch capabilities and sign URLs.

```
const capabilities = await ocsClient.getCapabilities()

const signedUrl = await ocsClient.signUrl('some_url_to_sign', 'your_username')
```

### WebDav

The webdav client needs to be instantiated with a base URI corresponding to your OpenCloud deployment. You can also pass a header callback which will be called with every dav request.

```
import { webdav } from '@opencloud-eu/web-client'

const accessToken = 'some_access_token'
const baseURI = 'some_base_uri'

const webDavClient = webdav(
	baseURI,
	() => ({ Authorization: accessToken })
)
```

The following example demonstrates how to list all resources of a given `SpaceResource` (see above how to fetch space resources).

```
const { resource, children } = await webDavClient.listFiles(spaceResource)
```
