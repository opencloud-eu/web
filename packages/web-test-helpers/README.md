# web-test-helpers

This packages provides utilities for unit testing within the OpenCloud app ecosystem.

## Installation

Depending on your package manager, run one of the following commands:

```
$ npm install @opencloud-eu/web-test-helpers --save-dev

$ pnpm add -D @opencloud-eu/web-test-helpers

$ yarn add @opencloud-eu/web-test-helpers --dev
```

Make sure that you have `vue`, `@vue/test-utils`, and `@opencloud-eu/web-pkg` as dev dependencies of your app since those are peer dependencies of this package.
