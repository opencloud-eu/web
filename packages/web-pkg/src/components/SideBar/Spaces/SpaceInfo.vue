<template>
  <div class="space_info grid items-center p-2">
    <div class="space_info__body flex items-center text-sm">
      <div class="mr-2">
        <oc-icon
          name="layout-grid"
          :size="resource.description ? 'large' : 'medium'"
          class="oc-display-block"
        />
      </div>
      <div>
        <h3
          data-testid="space-info-name"
          class="font-semibold m-0 text-base"
          v-text="resource.name"
        />
        <span data-testid="space-info-subtitle" v-text="resource.description" />
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent, inject } from 'vue'
import { SpaceResource } from '@opencloud-eu/web-client'

export default defineComponent({
  name: 'SpaceInfo',
  setup() {
    return {
      resource: inject<SpaceResource>('resource')
    }
  }
})
</script>

<style lang="scss">
.space_info {
  grid-template-columns: auto 1fr auto;
  grid-gap: 5px;
}
</style>
