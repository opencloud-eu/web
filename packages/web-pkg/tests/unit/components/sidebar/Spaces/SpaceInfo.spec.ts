import { SpaceResource } from '@opencloud-eu/web-client'
import SpaceInfo from '../../../../../src/components/SideBar/Spaces/SpaceInfo.vue'
import { defaultPlugins, shallowMount } from '@opencloud-eu/web-test-helpers'

const spaceMock = {
  type: 'space',
  name: ' space',
  id: '1',
  mdate: 'Wed, 21 Oct 2015 07:28:00 GMT',
  spaceQuota: {
    used: 100
  }
} as unknown as SpaceResource

const selectors = {
  name: '[data-testid="space-info-name"]',
  subtitle: '[data-testid="space-info-subtitle"]'
}

describe('SpaceInfo', () => {
  it('shows space info', () => {
    const { wrapper } = createWrapper(spaceMock)
    expect(wrapper.find(selectors.name).exists()).toBeTruthy()
    expect(wrapper.find(selectors.subtitle).exists()).toBeTruthy()
    expect(wrapper.html()).toMatchSnapshot()
  })
})

function createWrapper(spaceResource: SpaceResource) {
  return {
    wrapper: shallowMount(SpaceInfo, {
      global: {
        plugins: [...defaultPlugins()],
        provide: { resource: spaceResource }
      }
    })
  }
}
