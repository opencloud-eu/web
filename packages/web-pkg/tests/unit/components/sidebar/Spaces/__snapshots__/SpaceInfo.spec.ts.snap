// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SpaceInfo > shows space info 1`] = `
"<div class="space_info oc-p-s">
  <div class="space_info__body oc-text-overflow oc-flex oc-flex-middle">
    <div class="oc-mr-s">
      <oc-icon-stub accessiblelabel="" color="" filltype="fill" name="layout-grid" size="medium" type="span" class="oc-display-block"></oc-icon-stub>
    </div>
    <div>
      <h3 data-testid="space-info-name" class="oc-font-semibold oc-m-rm"> space</h3> <span data-testid="space-info-subtitle"></span>
    </div>
  </div>
</div>"
`;
